docs/ADR-016-Backwards-and-Forwards-Compatibility.md

```diff
@@ -0,0 +1,136 @@
+
+# ADR-016: Backwards and Forwards Compatibility
+
+**Status:** Proposed
+
+**Date:** 2025-10-07
+
+## Context
+
+We are working on [a new API](./ADR-004-Support-only-for-mainnet-and-upcoming-eras.md) for `cardano-api` that is centred around the `Era era` type. We would like to maintain backwards compatibility with the "old" api under two conditions:
+
+1) Existing exposed functionality that uses the "old" api which may be in use.
+2) Any functionality needed to enable QA's ability to test hardforking from Byron to the current era.
+
+## Proposed Solution
+
+The solution proposed below is specific to certificates but can be expanded to any era depdendent type.
+
+Our "old" Certificate definition is as follows:
+
+```haskell
+data Certificate era where
+  -- Pre-Conway
+  ShelleyRelatedCertificate
+    :: Typeable era
+    => ShelleyToBabbageEra era
+    -> Ledger.ShelleyTxCert (ShelleyLedgerEra era)
+    -> Certificate era
+  -- Conway onwards
+  ConwayCertificate
+    :: Typeable era
+    => ConwayEraOnwards era
+    -> Ledger.ConwayTxCert (ShelleyLedgerEra era)
+    -> Certificate era
+``` 
+
+`Cardano.Api.Certificate` exposes several helper functions that use this "old" type. E.g `makeStakeAddressDelegationCertificate`, `makeStakeAddressRegistrationCertificate` and `makeStakeAddressUnregistrationCertificate` to name a few. All of these "old" functions will be deprecated in favour of a new API that allows us to expose functionality from any era required without adding significant boilerplate as seen in the above definition of `data Certificate era`. We achieve this by the following:
+
+
+Replace era-variant GADTs (like Certificate era) with a uniform wrapper (Certificate :: L.TxCert era -> Certificate era) and using open type families to resolve era-specific types (e.g., Delegatee era). Era constraints (IsEra, IsShelleyBasedEra) gate access to functionality.
+
+I will make this clearer with an example below.
+
+**NB**: We will only implement functionality constrained by `IsShelleyBasedEra era` and `IsEra era` that either is currently exposed by the "old" api or specifically required by QA. If we do not need backwards compatibility then we will use`Era era` or `IsEra era` without implementing an additional type synonym family.
```

Comment:
Suggestion: use `[!NOTE]` admonition instead of NB https://github.com/orgs/community/discussions/16925 for better visual separation.